src/calculus.jl

```diff
@@ -0,0 +1,10 @@
+
+
+function diff(b1::Basic, b2::Basic)
```

Comment:
Does the `basic_get_type` function help? If it gives enough information, it
could be used to provide something like this:

```
function get_type(s::Basic)
    ccall((:basic_get_type, :libsymengine), Int, (Ptr{Basic},), &s)
end
## these typeIDs are a guess
type_map = Dict{Int, Function}(1 => BasicInteger, 2=>BasicRational,
11=>BasicSymbol)
```
Then you could return `type_map(get_type(x))(x)` from `add` and other
functions.

On Fri, Mar 25, 2016 at 3:27 PM, Isuru Fernando <notifications@github.com>
wrote:

> In src/calculus.jl
> <https://github.com/symengine/SymEngine.jl/pull/17#discussion_r57480885>:
>
> > @@ -0,0 +1,10 @@
> > +
> > +
> > +function diff(b1::Basic, b2::Basic)
>
> Yes, we want to promote everything into a Basic. That's basically what we
> do in C++, Python, etc.
>
> In SymEngine we use a type code system. It's an enum which gives the type
> of the basic object and we can use this to determine the output.
> For converting back this is useful.
>
> a = Basic()
> ccall(basic_add, Void, (Ptr{Basic}, Ptr{Basic}, Ptr{Basic}), &a, &b1, &b2)
> t = code to get enum value
> if t is Add
>     return Add(a)
> if t is Pow
>     return Pow(a)
> if t is Integer
>     return Integer(a)
>
> A SymEngine Integer can be converted to Julia Integers, but others can't
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/symengine/SymEngine.jl/pull/17/files/7343bd913def49c83b58a274d6363243ba1b9b57#r57480885>
>



-- 
John Verzani
Chair, Department of Mathematics
College of Staten Island, CUNY
verzani@math.csi.cuny.edu